src/server/context.ts

```diff
@@ -265,16 +269,75 @@ export class ServerContext {
         url.pathname = url.pathname.slice(0, -1);
         return Response.redirect(url.href, 307);
       }
-      const handle = () => Promise.resolve(inner(req, connInfo));
-      return middleware.handler(req, { handle, ...connInfo });
+      return withMiddlewares(req, connInfo, inner);
+    };
+  }
+
+  /**
+   * Identify which middlewares should be applied for a request,
+   * chain them and return a handler response
+   */
+  #composeMiddlewares(middlewares: MiddlewareRoute[]) {
+    return (req: Request, connInfo: ConnInfo, inner: router.Handler) => {
+      // identify middlewares to apply, if any, with their layer level
+      const mws = selectMiddlewares(req.url, middlewares);
+
+      // if no middleware, return handler
+      if (mws.length === 0) {
+        return inner(req, connInfo, {});
+      }
+
+      if (mws.length === 1) {
+        return mws[0].handler(req, {
+          handle: (state: Record<string, unknown> = {}) =>
+            Promise.resolve(inner(req, connInfo, state)),
+          ...connInfo,
+          state: {},
+        });
+      }
+
+      // apply all the middlewares
+      const deepestMw = mws.shift();
+      const shallowMw = mws.pop();
+      const handle0 = (outerState: Record<string, unknown> = {}) =>
+        Promise.resolve(
+          deepestMw!.handler(req, {
+            handle: (state: Record<string, unknown> = {}) =>
+              Promise.resolve(
+                inner(req, connInfo, { ...outerState, ...state }),
+              ),
+            ...connInfo,
+            state: { ...outerState },
+          }),
+        );
+
+      const handlers = [handle0];
+      for (const [i, mw] of mws.entries()) {
+        const handleItem = (outerState: Record<string, unknown> = {}) =>
+          Promise.resolve(
+            mw.handler(req, {
+              handle: (state: Record<string, unknown> = {}) =>
+                Promise.resolve(handlers[i]({ ...outerState, ...state })),
+              ...connInfo,
+              state: { ...outerState },
+            }),
+          );
+        handlers.push(handleItem);
+      }
+
+      return shallowMw!.handler(req, {
+        handle: handlers.pop()!,
+        ...connInfo,
+        state: {},
+      });
```

Comment:
This can be simplified a lot when using a mutable `ctx.state`:

```suggestion
      const handlers = [];

      const ctx = {
        next() {
          const handler = handlers.shift();
          return handler();
        },
        ...connInfo,
        state: {},
      }; 
     
      for (const mw of mws) {
        handlers.push(() => mw.handler(req, ctx));
      }
      
      // It'd be nice to not pass `connInfo` && `ctx.state` here, and instead just pass `ctx` directly.
      handlers.push(() => inner(req, connInfo, ctx.state));
      
      const handler = handlers.shift();
      return handler();
```